event-svc/src/event/validator/event.rs

```diff
@@ -0,0 +1,187 @@
+    pub invalid: Vec<InvalidItem>,
+}
+
+#[derive(Debug)]
+pub struct ValidatedEvent {
+    key: EventId,
+    cid: Cid,
+    event: Arc<unvalidated::Event<Ipld>>,
+}
+
+impl ValidatedEvent {
+    pub fn order_key(&self) -> &EventId {
+        &self.key
+    }
+}
+
+// TODO: should this impl from or have a private method?
+// it's not supposed to be public outside the crate and it currently allows us to
+// skip validation for the cases we're not doing it
+impl From<UnvalidatedEvent> for ValidatedEvent {
+    fn from(value: UnvalidatedEvent) -> Self {
+        Self {
+            key: value.key,
+            cid: value.cid,
+            event: value.event,
+        }
+    }
+}
+
+impl From<ValidatedEvent> for EventInsertable {
+    fn from(value: ValidatedEvent) -> Self {
+        EventInsertable::new(value.key, value.cid, value.event, false)
+            .expect("validated events must be insertable")
+    }
+}
+
+#[derive(Debug)]
+pub struct UnvalidatedEvent {
+    pub key: EventId,
+    pub cid: Cid,
+    pub event: Arc<unvalidated::Event<Ipld>>,
+}
+
+impl UnvalidatedEvent {
+    pub fn order_key(&self) -> &EventId {
+        &self.key
+    }
+}
+
+impl TryFrom<&ReconItem<EventId>> for UnvalidatedEvent {
+    type Error = crate::Error;
+
+    fn try_from(item: &ReconItem<EventId>) -> std::result::Result<Self, Self::Error> {
+        let (cid, event) = unvalidated::Event::<Ipld>::decode_car(item.value.as_slice(), false)
+            .map_err(Self::Error::new_app)?;
+
+        let key_cid = item.key.cid().ok_or_else(|| {
+            Self::Error::new_app(anyhow::anyhow!("EventId missing CID. EventID={}", item.key))
+        })?;
+
+        if key_cid != cid {
+            return Err(Self::Error::new_app(anyhow::anyhow!(
+                "EventId CID ({}) does not match the root CID of the CAR file ({})",
+                key_cid,
+                cid
+            )));
+        }
+
+        Ok(Self {
+            key: item.key.to_owned(),
+            cid: key_cid,
+            event: Arc::new(event),
+        })
+    }
+}
+
+impl ValidatedEvents {
+    pub fn new_with_expected_valid(valid: usize) -> Self {
+        // sort of arbitrary sizes, not betting on invalid events
+        Self {
+            valid: Vec::with_capacity(valid),
+            pending: Vec::with_capacity(valid / 4),
+            invalid: Vec::new(),
+        }
+    }
+
+    pub fn extend_with(&mut self, other: Self) {
+        self.valid.extend(other.valid);
+        self.invalid.extend(other.invalid);
+        self.pending.extend(other.pending);
+    }
+}
+
+#[derive(Debug)]
+pub struct EventValidator<'a> {
+    pool: &'a SqlitePool,
+    /// Whether to require the init event exists, and if we should check the the signature against
+    /// the init controller. True in all cases except a data migration.
+    require_init_event: bool,
+    /// Whether we should check the signature is currently valid or simply whether it was once valid
+    check_exp: bool,
```

Comment:
for recon we might get events that are old and we haven't yet decided to tie rpc time stamping in as a blocker for validating the event signatures (more below)... the API uses the wallclock time, but recon and migrations set this to false currently

Reply:
yeah, I think this comes back to what we started talking about at the end of the standup call earlier today.  Recon still needs to validate expiration times, it can't just totally ignore them.  I do think the logic for the api and recon likely should be the same and unified, I'm not sure we need this boolean.  But really we need to figure out how CACAO enforcement with time information from anchors is going to work.